### DISCUSSION TITLE
Adding V2xFullMessageReception

### DISCUSSION THREAD

**Discussion by dramsamooj2099** [eclipse/mosaic#292] [Q&A]
Hello,
I am trying to add the V2xFullMessageReception interaction to the output file (by default called visualizer.csv). I can't seem to replace the  V2xMessageReception interaction. I am trying to print out the message into the output file. Can I use V2xFullMessageReception or is there another way to print out the message?

Thank you

**Comment by schwepmo** [comment]
Hello,
this is probably due to the output ambassador not being subscribed to the `V2xFullMessageReception` by default.
Try adding `V2xFullMessageReception` to the subscriptions in the *runtime.json*-file located in the etc directory.

<pre>
        {
            "id": "output",
            "classname": "org.eclipse.mosaic.fed.output.ambassador.OutputAmbassador",
            "configuration": "output_config.xml",
            "priority": 50,
            "deploy": false,
            "start": false,
            "subscriptions": [
                "ChargingStationRegistration",
                "RsuRegistration",
                "TrafficLightRegistration",
                "TrafficSignRegistration",
                "VehicleRegistration",
                "ServerRegistration",
                "TmcRegistration",
                "ItefLogging",
                "VehicleLaneChange",
                "LanePropertyChange",
                "VehicleRouteChange",
                "TrafficLightStateChange",
                "TrafficSignSpeedLimitChange",
                "TrafficSignLaneAssignmentChange",
                "ChargingStationUpdate",
                "AdHocCommunicationConfiguration",
                "CellularCommunicationConfiguration",
                "V2xMessageRemoval",
                "V2xMessageReception",
                "V2xMessageTransmission",
                <b>"V2xFullMessageReception",</b>
                "EnvironmentSensorUpdates",
                "VehicleSlowDown",
                "TrafficDetectorUpdates",
                "TrafficLightUpdates",
                "VehicleBatteryUpdates",
                "CellularHandoverUpdates",
                "VehicleUpdates",
                "VehicleRoutesInitialization",
                "VehicleTypesInitialization"
            ],
            "javaClasspathEntries": []
        }
</pre>

**Reply by dramsamooj2099** [reply]
Hello,
I added V2xFullMessageReception to runtime.json file in the output federate and it still didn't work. Below is how I added it to my output_config.xml. Is there anything I need to do in my application for it to log V2xFullMessageReception?

Thank you

![output1](https://user-images.githubusercontent.com/99836337/219490998-6a9d205c-c684-477f-b140-e15e6adef574.PNG)
![output2](https://user-images.githubusercontent.com/99836337/219491000-dfcec45d-d07a-4dbb-b722-ab32de5907c6.PNG)

**Reply by schwepmo** [reply]
Hey it seems like you added the subscription to the `WebsocketVisualizer` in your `output_config.xml` but from what I can read in your question you want to add it to the file output (which you have done in the bottom picture?).

Have you validated that these interactions are actually send?

If you want you can add a zip of your scenario and I can try having a look at it.

**Reply by kschrab** [reply]
From what I know this specific interaction is currently not send from any of the pre-bundled simulators.

**Reply by dramsamooj2099** [reply]
Hello,
I attached my scenario and my applications running. I don't think the interaction (V2xFullMessageReception) is being sent but I don't know how to send it. Please let me know what you think,
[UMassDart.zip](https://github.com/eclipse/mosaic/files/10788559/UMassDart.zip)
[RSUComApp.zip](https://github.com/eclipse/mosaic/files/10788566/RSUComApp.zip)
[VehicleCommunicationApp.zip](https://github.com/eclipse/mosaic/files/10788567/VehicleCommunicationApp.zip)

**Reply by kschrab** [reply]
Currently, the actual V2xMessage object you pass to the `sendMessage` method is lost during communication simulation. This is due to the fact, that the communication simulator usually only requires the size of the message. This information is grabbed from the `EncodedMessage` object which is provided by the `V2xMessage`. All communication simulators we currently provide only use this field and omit the actual V2xMessage object and only inform the application simulator with the `V2xMessageReception` interaction, that a previously sent V2xMessage was received referred by the message id. 

The `V2xFullMessageReception` you've found was added to the codebase when we ran experiments where messages where generated by the communcation simulator itself, rather the application simulator, thus we needed a way to transmit the V2xMessage with the reception interaction. However, this code is currently not in the codebase thus the `V2xFullMessageReception` is not send at any time right now.

The better question would be, what exactly do you want to have in the output? Maybe post-processing the output.csv could help you with combining the message content from `V2xMessageTransmission` with the message id provided in `V2xMessageReception`?

As an alternative, you can of course adjust the codebase of MOSAIC in a way, that whenever a `V2xMessageReception` is instantiated, you create a `V2xFullMessageReception` instead.

**Reply by dramsamooj2099** [reply]
Hello,
Thanks for the additional information. My goal is to have one file that would be able to show the source and the destination with the contents of the message being sent and received. I thought I would be able to use V2xFullMessageReception to accomplish that but it seems I can't. Is there another way to accomplish this? Thanks for the help!

**Reply by kschrab** [reply]
Again, this is only possible if you adjust the source code of the communication simulator you use. Alternatively, you do that in a post-processing step with a self-written script (e.g. Pyhton) which runs over the output.csv and generates the missing information. 

Furthermore, looking at your scenario it seems you use SNS and NS-3 in parallel? Usually we would not recommend doing this as you would receive messages twice then.
